README.md
# NAME

Badge::Depot::App - Mojo app for hosting some badges

# VERSION

Version 0.0017, released 2016-08-12.

# SYNOPSIS

    $ cpanm Badge::Depot::App
    $ badge-depot-app initdb
    $ hypnotoad path/to/badge-depot-app

# DESCRIPTION

`badge-depot-app` is a [Mojolicious](https://metacpan.org/pod/Mojolicious) application for hosting a few [Badge::Depot](https://metacpan.org/pod/Badge::Depot) badges. An instance is running at
[https://badgedepot.code301.com](https://badgedepot.code301.com). It is only necessary to run this application if you wish to host your own badges.

It uses a SQLite database to cache information fetched from data sources.

# COMMANDS

## initdb

    $ badge-depot-app initdb

This command creates the data directory (by using `my_dist_data` in [File::HomeDir](https://metacpan.org/pod/File::HomeDir)) where the database and configuration
files are stored.

To re-create the database (for instance after an incompatible upgrade):

    $ badge-depot-app initdb --force

# HOSTED BADGES

The following badges are hosted by this application:

- [Badge::Depot::App::Plugin::Badge::Cpantesters](https://metacpan.org/pod/Badge::Depot::App::Plugin::Badge::Cpantesters)
- [Badge::Depot::App::Plugin::Badge::Kwalitee](https://metacpan.org/pod/Badge::Depot::App::Plugin::Badge::Kwalitee)

# SEE ALSO

- [Badge::Depot](https://metacpan.org/pod/Badge::Depot)
- [Task::Badge::Depot](https://metacpan.org/pod/Task::Badge::Depot)
- [Pod::Weaver::Section::Badges](https://metacpan.org/pod/Pod::Weaver::Section::Badges)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Badge-Depot-App

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.0017 2016-08-12T10:17:43Z UTC
 - No changes.

 [Stats & Dependency Changes]
 - 6 files changed, 9 insertions(+), 7 deletions(-)

0.0016 2016-08-12T09:37:38Z UTC
 [Bug Fixes]
 - Render old kwalitee badges (without author in url) as 'unknown', rather
   than not rendering at all.

 [Stats & Dependency Changes]
 - 6 files changed, 15 insertions(+), 9 deletions(-)
 - (dev req) + Dist::Zilla::Plugin::PromptIfStale (any)
 - (dev sug) ~ Dist::Iller::Config::Author::CSSON 0.0310 --> 0.0311

0.0015 2016-08-11T21:45:09Z UTC
 [Bug Fixes]
 - Fixed Kwalitee badge by adapting to cpants changes (url now contains
   'author').

 [Stats & Dependency Changes]
 - 7 files changed, 24 insertions(+), 19 deletions(-)
 - (run req) ~ DateTime (any) --> 1.05
 - (run req) ~ List::Util (any) --> 1.45
 - (run req) ~ Path::Tiny (any) --> 0.072
 - (run req) ~ Safe::Isa (any) --> 1.000005
 - (run req) ~ Try::Tiny (any) --> 0.24
 - (run req) ~ version (any) --> 0.9913
 - (dev req) + Dist::Zilla 5.000
 - (dev req) + Dist::Zilla::Plugin::Git 2.036
 - (dev req) + Dist::Zilla::Plugin::MetaConfig (any)
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::Version 1
 - (dev req) ~ Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   (any) --> 0.007
 - (dev req) ~ Dist::Zilla::Plugin::ChangeStats::Dependencies::Git (any)
   --> 0.0100
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::Clean (any) --> 0.07
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild (any) --> 0.151680
 - (dev req) ~ Dist::Zilla::Plugin::Git::Contributors (any) --> 0.017
 - (dev req) ~ Dist::Zilla::Plugin::InstallRelease (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package (any) -->
   2.003001
 - (dev req) ~ Dist::Zilla::Plugin::NextRelease::Grouped (any) --> 0.0200
 - (dev req) ~ Dist::Zilla::Plugin::PodWeaver (any) --> 4.006
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod (any) --> 0.150250
 - (dev req) ~ Dist::Zilla::Plugin::RunExtraTests (any) --> 0.028
 - (dev req) ~ Dist::Zilla::Plugin::Test::Compile (any) --> 2.054
 - (dev req) ~ Dist::Zilla::Plugin::Test::EOF (any) --> 0.0501
 - (dev req) ~ Dist::Zilla::Plugin::Test::EOL (any) --> 0.18
 - (dev req) ~ Dist::Zilla::Plugin::Test::Kwalitee::Extra (any) --> v0.2.1
 - (dev req) ~ Dist::Zilla::Plugin::Test::NoTabs (any) --> 0.15
 - (dev req) ~ Dist::Zilla::Plugin::Test::ReportPrereqs (any) --> 0.021
 - (dev req) ~ Dist::Zilla::Plugin::Test::Version (any) --> 1.05
 - (dev req) ~ Dist::Zilla::Plugin::TravisYML (any) --> 1.15
 - (dev req) ~ Dist::Zilla::Plugin::UploadToStratopan (any) --> 0.012
 - (dev req) ~ Test::NoTabs (any) --> 1.4
 - (dev req) ~ Test::Warnings (any) --> 0.026
 - (dev req) - Dist::Zilla::Plugin::MetaProvides::Class
 - (dev req) - Dist::Zilla::Plugin::PodnameFromClassname
 - (dev req) - Dist::Zilla::Plugin::Prereqs::Plugins
 - (dev req) - Pod::Weaver::Section::Homepage::DefaultCPAN
 - (dev req) - Pod::Weaver::Section::Source::DefaultGitHub
 - (dev sug) ~ Dist::Iller 0.1406 --> 0.1408
 - (dev sug) ~ Dist::Iller::Config::Author::CSSON 0.0306 --> 0.0310

0.0014 2016-02-17T12:52:32Z UTC
 [Documentation]
 - Added some documentation.

 [Stats & Dependency Changes]
 - 8 files changed, 192 insertions(+), 132 deletions(-)
 - (dev req) - Pod::Weaver::Section::Badges

0.0013 2016-02-17T01:07:30Z UTC
 [Bug Fixes]
 - Fixed badge mime type.

 [Stats & Dependency Changes]
 - 6 files changed, 13 insertions(+), 10 deletions(-)

0.0012 2016-02-17T00:43:57Z UTC
 [Bug Fixes]
 - Fix CpanTesters badge if we don't have a json object.

 [Stats & Dependency Changes]
 - 6 files changed, 10 insertions(+), 7 deletions(-)
 - (dev req) - Dist::Iller
 - (dev req) - Dist::Iller::Config::Author::CSSON
 - (dev sug) + Dist::Iller 0.1406
 - (dev sug) + Dist::Iller::Config::Author::CSSON 0.0306

0.0011 2016-02-16T22:41:10Z UTC
 [New Features]
 - Added Kwalitee badge.

 [Stats & Dependency Changes]
 - 6 files changed, 101 insertions(+), 16 deletions(-)

0.0010 2016-02-16T19:43:02Z UTC
 [Enhancements]
 - Spew a string with a hash, not json.

 [Stats & Dependency Changes]
 - 5 files changed, 5 insertions(+), 6 deletions(-)

0.0009 2016-02-16T19:40:14Z UTC
 [Bug Fixes]
 - Spew json, not a hash.

 [Stats & Dependency Changes]
 - 5 files changed, 7 insertions(+), 6 deletions(-)

0.0008 2016-02-16T19:37:39Z UTC
 [Enhancements]
 - Better data dir creation.

 [Stats & Dependency Changes]
 - 5 files changed, 7 insertions(+), 6 deletions(-)

0.0007 2016-02-16T19:30:05Z UTC
 [Bug Fixes]
 - Call String::Random constructor

 [Stats & Dependency Changes]
 - 5 files changed, 6 insertions(+), 6 deletions(-)

0.0006 2016-02-16T19:28:18Z UTC
 [Bug Fixes]
 - Make it possible to run 'initdb'.

 [Stats & Dependency Changes]
 - 5 files changed, 12 insertions(+), 8 deletions(-)

0.0005 2016-02-16T19:23:51Z UTC
 [Bug Fixes]
 - Make it possible to run 'initdb'.

 [Stats & Dependency Changes]
 - 5 files changed, 14 insertions(+), 8 deletions(-)

0.0004 2016-02-16T19:20:03Z UTC
 [Enhancements]
 - Better message on first run (data dir).

 [Stats & Dependency Changes]
 - 5 files changed, 7 insertions(+), 7 deletions(-)

0.0003 2016-02-16T19:18:02Z UTC
 [Enhancements]
 - Better message on first run (data dir).

 [Stats & Dependency Changes]
 - 5 files changed, 12 insertions(+), 6 deletions(-)

0.0002 2016-02-16T19:12:59Z UTC
 - Initial release


lib/Badge/Depot/App.pm
use 5.10.0;
use strict;
use warnings;

package Badge::Depot::App;

# ABSTRACT: Mojo app for hosting some badges
# AUTHORITY
our $VERSION = '0.0018';

1;

__END__

=pod

=head1 SYNOPSIS

    $ cpanm Badge::Depot::App
    $ badge-depot-app initdb
    $ hypnotoad path/to/badge-depot-app

=head1 DESCRIPTION

C<badge-depot-app> is a L<Mojolicious> application for hosting a few L<Badge::Depot> badges. An instance is running at
L<https://badgedepot.code301.com>. It is only necessary to run this application if you wish to host your own badges.

It uses a SQLite database to cache information fetched from data sources.

=head1 COMMANDS

=head2 initdb

    $ badge-depot-app initdb

This command creates the data directory (by using C<my_dist_data> in L<File::HomeDir>) where the database and configuration
files are stored.

To re-create the database (for instance after an incompatible upgrade):

    $ badge-depot-app initdb --force

=head1 HOSTED BADGES

The following badges are hosted by this application:

=for :list
* L<Badge::Depot::App::Plugin::Badge::Cpantesters>
* L<Badge::Depot::App::Plugin::Badge::Kwalitee>

=head1 SEE ALSO

=for :list
* L<Badge::Depot>
* L<Task::Badge::Depot>
* L<Pod::Weaver::Section::Badges>

=cut


lib/Badge/Depot/App/Plugin/DB.pm
use 5.10.0;
use strict;
use warnings;

package Badge::Depot::App::Plugin::DB;

# AUTHORITY
our $VERSION = '0.0018';

use Mojo::Base 'Mojolicious::Plugin';
use File::HomeDir qw/my_dist_data/;
use Mojo::SQLite;
use Path::Tiny;
use Try::Tiny;
use DateTime;
use List::Util qw/any/;

has 'sqlite';
has 'db';

sub register {
    my $self = shift;
    my $app = shift;
    my $conf = shift;


    my $data_dir;
    try {
        $data_dir = path(my_dist_data('Badge-Depot-App'));
    }
    catch {};

    if(!$data_dir) {
        warn "It appears you have no data directory. Run 'badge-depot-app initdb' to set it up";
        return;
    }

    my $dbfile = $data_dir->child('badge_depot_app.db');
    if(!$dbfile->exists) {
        warn "It appears you have no database. Run 'badge-depot-app initdb' to set it up";
        return;
    }

    $self->sqlite(Mojo::SQLite->new('sqlite:' . $dbfile->realpath));
    $self->db($self->sqlite->db);
    $app->helper(db => sub { $self });

    return $self;
}

sub get_value {
    my $self = shift;
    my $badge = shift;

    my $info = shift;
    my $init_expiry = $info->{'init_expiry'} || 'hours';
    $init_expiry = (any { $init_expiry eq $_ } qw/hours days weeks months/) ? $init_expiry : 'hours';
    my $dist = $info->{'dist'};
    my $version = $info->{'version'};
    my $current = $info->{'current'};

    my $now = DateTime->now->strftime('%Y-%m-%d %H:%M:%S');

    my $results = $self->db->query(q{
        SELECT value, color, earliest_change_at, change_unit
          FROM distribution
         WHERE badge = ?
           AND distname = ?
           AND version = ?}, $badge, $dist, $version);

    my $row = $results->hash;
    if($row) {
        # don't check yet
        if($row->{'earliest_change_at'} gt $now) {
            return { value => $row->{'value'}, color => $row->{'color'} };
        }
        # check current
        else {
            my $refreshed = $current->($info);

            if(!defined $refreshed) {
                $refreshed = { value => 'unknown', color => 'lightgrey' };
            }

            my $value = $refreshed->{'value'};
            my $color = $refreshed->{'color'};

            my($earliest_change_at, $change_unit) = $self->get_earliest_change_at($row->{'value'}, $value, $row->{'change_unit'});
            $self->db->query(q{
                UPDATE distribution
                   SET value = ?,
                       color = ?,
                       changed_at = ?,
                       earliest_change_at = ?,
                       change_unit = ?
                 WHERE distname = ?
                   AND version = ?
                   AND badge = ?
                }, $value, $color, $now, $earliest_change_at, $change_unit, $dist, $version, $badge);
            return { value => $row->{'value'}, color => $row->{'color'} };
        }
    }
    # new dist+version+badge
    else {
        my $refreshed = $current->($info);

        if(!defined $refreshed) {
            $refreshed = { value => 'unknown', color => 'lightgrey' };
        }

        my $value = $refreshed->{'value'};
        my $color = $refreshed->{'color'};

        my($earliest_change_at, $change_unit) = $self->get_earliest_change_at(undef, $value, $init_expiry);

        $self->db->query(q{
            INSERT INTO distribution (distname, version, badge, value, color, created_at, changed_at, earliest_change_at, change_unit) values (?, ?, ?, ?, ?, ?, ?, ?, ?)},
                                      $dist, $version, $badge, $value, $color, $now, $now, $earliest_change_at, $change_unit
        );
        return { value => $value, color => $color };
    }
}

sub get_earliest_change_at {
    my $self = shift;
    my $previous_value = shift;
    my $current_value = shift;
    my $change_unit = shift;

    my @expiries = qw/hours days weeks months/;
    my($current_expiry_index) = grep { $expiries[$_] eq $change_unit } 0..$#expiries;
    my $next_expiry = $current_expiry_index == $#expiries ? $change_unit : $expiries[ $current_expiry_index + 1];

    my $actual_unit = !defined $previous_value          ? 'hours'
                    : !defined $current_value           ? 'hours'
                    : $previous_value eq $current_value ? $next_expiry
                    :                                     $change_unit
                    ;

    my $earliest_change_at = DateTime->now->add(
        $actual_unit eq 'hours'  ? (hours => 12)
      : $actual_unit eq 'days'   ? (days => 2)
      : $actual_unit eq 'weeks'  ? (weeks => 2)
      : $actual_unit eq 'months' ? (months => 3)
      :                            (days => 2)
    )->strftime('%Y-%m-%d %H:%M:%S');

    return ($earliest_change_at, $actual_unit);

}

# pass it on
sub query {
    shift->db->query(@_);
}

1;

__END__


lib/Badge/Depot/App/Command/initdb.pm
use 5.10.0;
use strict;
use warnings;

package Badge::Depot::App::Command::initdb;

# AUTHORITY
our $VERSION = '0.0018';

use Mojo::Base 'Mojolicious::Command';
use Getopt::Long qw/GetOptionsFromArray/;
use File::HomeDir qw/my_dist_data/;
use Path::Tiny;
use Try::Tiny;
use Mojo::SQLite;
use String::Random;

sub run {
    my $self = shift;
    my $args = \@_;

    my $force = 0;
    GetOptionsFromArray $args, 'f|force' => \$force;

    my $data_dir;
    try {
        $data_dir = path(my_dist_data('Badge-Depot-App'));
    }
    catch {
        my_dist_data('Badge-Depot-App', { create => 1 });
        $data_dir = path(my_dist_data('Badge-Depot-App'));
        $data_dir->child('mojo.secrets')->spew(String::Random->new->randpattern('c' x 40));
        $data_dir->child('mojo.conf')->spew('{ hypnotoad => {} }');
        say "Created data directory for Badge-Depot-App";
    };

    my $db = $data_dir->child('badge_depot_app.db');

    if($db->exists && !$force) {
        die sprintf 'The database exists at %s: Use --force to overwrite', $db->realpath;
    }
    if($db->exists) {
        $db->remove;
        say sprintf 'Has removed %s', $db->realpath;
    }

    my $sqlite = Mojo::SQLite->new('sqlite:' . $db->realpath);

    $sqlite->db->query(q{
        CREATE TABLE distribution (
            dist_id INTEGER PRIMARY KEY AUTOINCREMENT,
            distname TEXT NOT NULL,
            version TEXT NOT NULL,
            badge TEXT NOT NULL,
            value TEXT NOT NULL,
            color TEXT NOT NULL,
            created_at TEXT NOT NULL,
            changed_at TEXT NOT NULL,
            earliest_change_at TEXT NOT NULL,
            change_unit TEXT NOT NULL
        )
    });
    $sqlite->db->query(q{
        CREATE TABLE shield (
            badge_url TEXT NOT NULL PRIMARY KEY,
            badge_svg TEXT NOT NULL
        )
    });
    say sprintf 'Has created %s', $db->realpath;

}

1;

__END__


lib/Badge/Depot/App/Plugin/Badge/Kwalitee.pm
use 5.10.0;
use strict;
use warnings;

package Badge::Depot::App::Plugin::Badge::Kwalitee;

# AUTHORITY
our $VERSION = '0.0018';

use Mojo::Base 'Mojolicious::Plugin';
use Mojo::JSON qw/encode_json/;
use Mojo::UserAgent;

sub register {
    my $self = shift;
    my $app = shift;
    my $conf = shift;

    $app->routes->get('/badge/kwalitee/:author/:dist/*version', sub {
        my $c = shift;
        (my $dist = $c->param('dist')) =~ s{::}{-}g;
        my $version = $c->param('version');
        my $author = $c->param('author');

        my $current = $app->db->get_value('kwalitee', {
            init_expiry => 'hours',
            dist => $dist,
            version => $version,
            author => $author,
            current => $self->current,
        });

        $app->render_badge($c, 'kwalitee', $current->{'value'}, $current->{'color'});

    });

    # keep the old url for now, but render 'unknown' for all
    $app->routes->get('/badge/kwalitee/:dist/*version', sub {
        my $c = shift;
        $app->render_badge($c, 'kwalitee', 'unknown', 'lightgrey');
    });

    return $app;
}

sub current {
    my $self = shift;

    return sub {
        my $info = shift;
        my $dist = $info->{'dist'};
        my $version = $info->{'version'};
        my $author = $info->{'author'};

        my $tx = Mojo::UserAgent->new->get("http://cpants.cpanauthors.org/release/$author/$dist-$version.json");

        if(!$tx->success) {
            return;
        }

        my $data = $tx->res->json;
        my $core_kwalitee = $data->{'distribution'}{'core_kwalitee'};
        my $kwalitee = $data->{'distribution'}{'kwalitee'};

        my $color = $core_kwalitee == 100 && $kwalitee > 130   ? 'brightgreen'
                  : $kwalitee >= 128                           ? 'green'
                  : $kwalitee >= 120                           ? 'yellow'
                  : $kwalitee >= 110                           ? 'orange'
                  :                                              'red'
                  ;
        return { value => sprintf ('%.2f' => $kwalitee), color => $color };
    };
}

1;

__END__


lib/Badge/Depot/App/Plugin/Badge/CpanTesters.pm
use 5.10.0;
use strict;
use warnings;

package Badge::Depot::App::Plugin::Badge::CpanTesters;

# AUTHORITY
our $VERSION = '0.0014';

use Mojo::Base 'Mojolicious::Plugin';
use Mojo::JSON qw/encode_json/;
use Mojo::UserAgent;
use Try::Tiny;
use CPAN::Testers::WWW::Reports::Parser;
use version;
use Safe::Isa qw/$_isa/;
use List::Util qw/sum/;
use URL::Encode qw/url_encode/;

sub register {
    my $self = shift;
    my $app = shift;
    my $conf = shift;

    return $app->routes->get('/badge/cpantesters/:dist/*version' => sub {
        my $c = shift;
        (my $dist = $c->param('dist')) =~ s{::}{-}g;
        my $version = $c->param('version');

        my $current = $app->db->get_value('cpantesters', {
            init_expiry => 'hours',
            dist => $dist,
            version => $version,
            current => $self->current,
        });

        $app->render_badge($c, 'cpantesters', $current->{'value'}, $current->{'color'});
    });

    return $app;
}

sub current {
    my $self = shift;

    return sub {
        my $dist = shift;
        my $version = shift;
        my $first_letter = substr $dist, 0, 1;

        my $json = Mojo::UserAgent->new->get("http://www.cpantesters.org/distro/$first_letter/$dist.json")->res->json;
        return if !defined $json;

        my $report;
        try {
            $report = CPAN::Testers::WWW::Reports::Parser->new(
                format => 'JSON',
                objects => 1,
                data => encode_json($json),
            );
        }
        catch { };
        return if !defined $report;

        if($version eq 'latest') {
            my $latest_version;
            while(my $data = $report->report) {
                my $report_version = version->new($data->version);

                $latest_version = $report_version if !defined $latest_version;
                $latest_version = $report_version if $report_version > $latest_version;
            }
            $version = $latest_version;
        }
        return if !defined $version || $version eq 'latest';
        $version = version->new($version) if !$version->$_isa('version');

        my %stats = (
            PASS => 0,
            NA => 0,
            UNKNOWN => 0,
            INVALID => 0,
            FAIL => 0,
        );

        while(my $data = $report->report) {
            next if version->parse($data->version) != $version;
            ++$stats{ $data->status };
        }

        return if sum(values %stats) == 0;

        my $ok = $stats{'PASS'};
        my $fails = $stats{'FAIL'};
        my $na = $stats{'NA'};
        my $unknowns = $stats{'NA'} + $stats{'UNKNOWN'} + $stats{'INVALID'};

        # NA doesn't count towards the color
        my $percent = 100 * $ok / ($ok + $fails + $stats{'UNKNOWN'} + $stats{'INVALID'});

        my $color = $percent >= 99   ? 'brightgreen'
                  : $percent >= 97.5 ? 'green'
                  : $percent >= 95   ? 'yellowgreen'
                  : $percent >= 90   ? 'yellow'
                  : $percent >= 80   ? 'orange'
                  :                    'red'
                  ;

        return { value => "$ok / $unknowns / $fails", color => $color };

    };
}

1;

__END__


lib/Badge/Depot/App/Plugin/Badge/Cpantesters.pm
use 5.10.0;
use strict;
use warnings;

package Badge::Depot::App::Plugin::Badge::Cpantesters;

# AUTHORITY
our $VERSION = '0.0018';

use Mojo::Base 'Mojolicious::Plugin';
use Mojo::JSON qw/encode_json/;
use Mojo::UserAgent;
use Try::Tiny;
use CPAN::Testers::WWW::Reports::Parser;
use version;
use Safe::Isa qw/$_isa/;
use List::Util qw/sum/;
use URL::Encode qw/url_encode/;

sub register {
    my $self = shift;
    my $app = shift;
    my $conf = shift;

    $app->routes->get('/badge/cpantesters/:dist/*version', sub {
        my $c = shift;
        (my $dist = $c->param('dist')) =~ s{::}{-}g;
        my $version = $c->param('version');

        my $current = $app->db->get_value('cpantesters', {
            init_expiry => 'hours',
            dist => $dist,
            version => $version,
            current => $self->current,
        });

        $app->render_badge($c, 'cpantesters', $current->{'value'}, $current->{'color'});
    });

    return $app;
}

sub current {
    my $self = shift;

    return sub {
        my $info = shift;
        my $dist = $info->{'dist'};
        my $version = $info->{'version'};
        my $first_letter = substr $dist, 0, 1;

        my $json = Mojo::UserAgent->new->get("http://www.cpantesters.org/distro/$first_letter/$dist.json")->res->json;
        return if !defined $json;

        my $report;
        try {
            $report = CPAN::Testers::WWW::Reports::Parser->new(
                format => 'JSON',
                objects => 1,
                data => encode_json($json),
            );
        }
        catch { };
        return if !defined $report;

        if($version eq 'latest') {
            my $latest_version;
            while(my $data = $report->report) {
                my $report_version = version->new($data->version);

                $latest_version = $report_version if !defined $latest_version;
                $latest_version = $report_version if $report_version > $latest_version;
            }
            $version = $latest_version;
        }
        return if !defined $version || $version eq 'latest';
        $version = version->new($version) if !$version->$_isa('version');

        my %stats = (
            PASS => 0,
            NA => 0,
            UNKNOWN => 0,
            INVALID => 0,
            FAIL => 0,
        );

        while(my $data = $report->report) {
            next if version->parse($data->version) != $version;
            ++$stats{ $data->status };
        }

        return if sum(values %stats) == 0;

        my $ok = $stats{'PASS'};
        my $fails = $stats{'FAIL'};
        my $na = $stats{'NA'};
        my $unknowns = $stats{'NA'} + $stats{'UNKNOWN'} + $stats{'INVALID'};

        # NA doesn't count towards the color
        my $percent = 100 * $ok / ($ok + $fails + $stats{'UNKNOWN'} + $stats{'INVALID'});

        my $color = $percent >= 99   ? 'brightgreen'
                  : $percent >= 97.5 ? 'green'
                  : $percent >= 95   ? 'yellowgreen'
                  : $percent >= 90   ? 'yellow'
                  : $percent >= 80   ? 'orange'
                  :                    'red'
                  ;

        return { value => "$ok / $unknowns / $fails", color => $color };

    };
}

1;

__END__


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.007.
use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Mojo app for hosting some badges",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Badge-Depot-App",
  "EXE_FILES" => [
    "script/badge-depot-app"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "Badge::Depot::App",
  "PREREQ_PM" => {
    "CPAN::Testers::WWW::Reports::Parser" => 0,
    "DateTime" => "1.05",
    "File::HomeDir" => 0,
    "Getopt::Long" => 0,
    "List::Util" => "1.45",
    "Module::Find" => 0,
    "Mojo::SQLite" => 0,
    "Mojolicious" => "6.00",
    "Path::Tiny" => "0.072",
    "Safe::Isa" => "1.000005",
    "String::Random" => 0,
    "Try::Tiny" => "0.24",
    "URL::Encode" => 0,
    "version" => "0.9913"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.0018",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "CPAN::Testers::WWW::Reports::Parser" => 0,
  "DateTime" => "1.05",
  "ExtUtils::MakeMaker" => 0,
  "File::HomeDir" => 0,
  "File::Spec" => 0,
  "Getopt::Long" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "List::Util" => "1.45",
  "Module::Find" => 0,
  "Mojo::SQLite" => 0,
  "Mojolicious" => "6.00",
  "Path::Tiny" => "0.072",
  "Safe::Isa" => "1.000005",
  "String::Random" => 0,
  "Test::More" => "0.96",
  "Try::Tiny" => "0.24",
  "URL::Encode" => 0,
  "version" => "0.9913"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
